src/libraries/Microsoft.PowerFx.Core/Public/Values/CollectionTableValue.cs

```diff
@@ -60,10 +89,30 @@ public override IEnumerable<DValue<RecordValue>> Rows
         {
             get
             {
-                foreach (var item in _enumerator)
+                if (_enumerator != null)
+                {
+                    foreach (var item in _enumerator)
+                    {
+                        var record = Marshal(item);
+                        yield return record;
+                    }
+                }
+                else
                 {
-                    var record = Marshal(item);
-                    yield return record;
+                    IAsyncEnumerator<T> asyncEnumerator = _asyncEnumerable.GetAsyncEnumerator();
+
+                    try
+                    {                        
+                        while (asyncEnumerator.MoveNextAsync().GetAwaiter().GetResult())
```

Comment:
Can we use foreach?

Reply:
we cant' use 'await foreach' here as Rows isn't async

Reply:
But can we factor it into a generic helper:

IEnum<T> Sync(IAsyncEnum<T> inner);

We're calling GetResult() either way and doing sync-over-async.

That would quarantine the code. 
And then this method can just use Foreach on the Sync() helper.